pkg/controller/component/component_controller.go

```diff
@@ -0,0 +1,182 @@
+	cmpopts.IgnoreFields(corev1.ConfigMap{}, "TypeMeta", "ObjectMeta"),
+}
+
+// Add creates a new Component Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileComponent{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("component-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource Component
+	err = c.Watch(&source.Kind{Type: &workspacev1alpha1.Component{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(&source.Kind{Type: &corev1.ConfigMap{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &workspacev1alpha1.Component{},
+	})
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileComponent implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileComponent{}
+
+// ReconcileComponent reconciles a Component object
+type ReconcileComponent struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a Component object and makes changes based on the state read
+// and what is in the Component.Spec
+func (r *ReconcileComponent) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling Component")
+
+	// Fetch the Component instance
+	instance := &workspacev1alpha1.Component{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	var components []workspacev1alpha1.ComponentDescription
+	dockerimageDevfileComponents, pluginDevfileComponents, err := adaptor.SortComponentsByType(instance.Spec.Components)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	commands := instance.Spec.Commands
+
+	dockerimageComponents, err := adaptor.AdaptDockerimageComponents(instance.Spec.WorkspaceId, dockerimageDevfileComponents, commands)
```

Comment:
In general dockerimage is no a required component, so additional check is needed to avoid 
```
{"level":"error","ts":1586176842.3509479,"logger":"controller-runtime.controller","msg":"Reconciler error","controller":"workspace-controller","request":"dev-ws/cloud-shell","error":"Component.workspace.che.eclipse.org \"components-workspaced46b62f0ebe44f3c-docker\" is invalid: spec.components: Invalid value: \"null\": spec.components in body must be of type array: \"null\"","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.3.0/pkg/internal/controller/controller.go:218\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.3.0/pkg/internal/controller/controller.go:192\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.3.0/pkg/internal/controller/controller.go:171\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:88"}
```
https://github.com/che-incubator/che-workspace-operator/commit/64751c6ee3ead5790b1b78288458fe5ac4f866d9
tested with theia-simple https://github.com/che-incubator/che-workspace-operator/commit/a0b4bb05e4adeb78f6178e834f6e5ce794c65203
P.S. I was lucky not being able to start any other workspace with plugins and dockerimage containers because of memory lack and it's how I found this bug.

P.P.S. Not sure if we should be able to run workspace without editor/plugins, like with dockerimage component. In general it could work, so:
- we should add a corresponding check below;
- add some devfile validating before workspace creating;

Reply:
Good catch -- I had completely missed this. I'm adding checks to `pkg/controller/workspace/provision/component.go` to avoid this. For now, I'm allowing workspaces with no editors/plugins (though they're kind of meaningless). We can decide separately if we e.g. want a default as is in upstream or if we want to just set status to failed. 

As for devfile validation, I expect that will come once we move to the devfile 2.0 spec, as validation will be built in.